File: solutions/315.count-of-smaller-numbers-after-self/count-of-smaller-numbers-after-self.py
class Solution(object):
    def countSmaller(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        if not nums:
            return []
        mx = max(nums)
        mi = min(nums)

        res = [0] * len(nums)
        idx = len(nums) - 1
        tree = [0] * (mx - mi + 2)
        for i in range(len(nums)-1, -1, -1):
            n = nums[i]
            res[idx] = self.getCount(n - mi, tree)
            idx -= 1
            self.update(n - mi+1, tree)
        return res

    def getCount(self, n, tree):
        cnt = 0
        while n > 0:
            cnt += tree[n]
            n -= n & (-n)
        return cnt

    def update(self, n, tree):
        while n < len(tree):
            tree[n] += 1
            n += n & (-n)

# so = Solution()
# print(so.countSmaller([5,2,1,6]))

File: solutions/315.count-of-smaller-numbers-after-self/count-of-smaller-numbers-after-self.java
class Solution {
    public List<Integer> countSmaller(int[] nums) {
        int n = nums.length;
        int min = 0, max = 0;
        for (int i = 0; i < n; i++) {
            min = Math.min(nums[i], min);
            max = Math.max(nums[i], max);
        }
        int[] tree = new int[max - min + 2];
        List<Integer> ans = new ArrayList();

        for (int i = n - 1; i >= 0; i--) {
            int x = get(tree, nums[i] - min);
            ans.add(x);
            update(tree, nums[i] - min + 1);
        }

        Collections.reverse(ans);
        return ans;
    }

    public void update(int[] tree, int n) {
        for (int i = n; i < tree.length; i += i & (-i)) {
            tree[i] += 1;
        }
    }

    public int get(int[] tree, int n) {
        int sum = 0;
        for (int i = n; i > 0; i -= i & (-i)) {
            sum += tree[i];
        }
        return sum;
    }
}